onnxruntime/core/dll/dllmain.cc

```diff
@@ -40,7 +46,14 @@ BOOL APIENTRY DllMain(HMODULE /*hModule*/,
   switch (ul_reason_for_call) {
     case DLL_PROCESS_ATTACH:
     case DLL_THREAD_ATTACH:
+      break;
     case DLL_THREAD_DETACH:
+#if defined(_WIN32) && defined(USE_CUDA)
```

Comment:
If  ul_reason_for_call is DLL_PROCESS_DETACH, lpvReserved is NULL if FreeLibrary has been called or the DLL load failed and non-NULL if the process is terminating.

So what we can do there is: ul_reason_for_call is DLL_PROCESS_DETACH and the lpvReserved is not NULL, set g_cuda_detached = true. In this case, skip all the cleanups.